Migrate the bm_diff benchmarks to python3

drfloob opened the pull request:
Includes a requirements.txt pinned at ~2017 versions, when this script was first written.

lidizheng reviewed (commented):
The change looks pretty good. But the CI jobs seems unhappy.

drfloob commented on tools/profiling/microbenchmarks/bm_diff/requirements.txt:
```diff
@@ -0,0 +1,4 @@
+tabulate>=0.8
+scipy>=1.0.0
+requests>=2.19.0
+jwt>=0.5.0
```
No, the tools/gce/linux_kokoro_performance_worker_init.sh file installs dependencies for CI. This file is for human convenience, since I had to figure out what these requirements were myself when I wanted to run the benchmarks locally.

drfloob replied:
Correction: I liked that idea, and changed the test setup script to utilize this requirements file. There's some definite redundancy in our python setup scripts though, I had to change the cryptography constraints in 3 different files for example. It's not trivial to deduplicate them, so I think that should be a cleanup for another time as well.

drfloob commented:
> The change looks pretty good. But the CI jobs seems unhappy.

Yes, sorry. It should have been a draft PR, I converted it to one. I'm still tracking down all the python2.7 dependencies related to this change, I'm not sure the scope of work yet.

drfloob commented:
This is ready for review, all tests had passed (I pushed and reverted an experiment).

I added cryptography version constraints here due to a jwt dependency update (jpadilla/pyjwt/pull/523), which is what caused our recent headache with the iOS binary size diff test (see #25053). Using a recent pyjwt library makes life easier for py2 and py3 coexistence since pyjwt v2.x has python3 return strings instead of bytes from the `encode` method, same as python2's behavior.

drfloob commented on tools/profiling/microbenchmarks/bm_diff/bm_diff.py:
```diff
@@ -91,7 +90,7 @@ def _args():
 
 def _maybe_print(str):
```
I like that idea, it's definitely a benefit we get from migrating to python3. Do you mind if it gets handled in another PR? I'd like to do this one change at a time.

lidizheng commented on tools/internal_ci/helper_scripts/prepare_build_linux_perf_rc:
```diff
@@ -19,9 +19,10 @@
 ulimit -c unlimited
 
-sudo pip install tabulate
+python3 -m pip install --upgrade pip
```
nit: let's use a pinned version of pip `19.3.1`. Bad release of pip broke us before.

drfloob replied:
Good catch, thanks. I wasn't aware of that.

lidizheng commented on tools/internal_ci/helper_scripts/prepare_build_macos_rc:
```diff
@@ -67,7 +67,8 @@ then
   # Needed for ios-binary-size
-  time pip install --user pyyaml pyjwt==1.7.1 pyOpenSSL cryptography requests
+  time pip install --user pyyaml pyjwt>=2.0.1 pyOpenSSL \
+    'cryptography>=3.3.1,<4.0.0' requests
```
optional: if the package requirement is complex, maybe we want to put them into a requirements.txt? Just like what you did for Linux.

drfloob replied:
Done, mostly. I deduplicated one overlapping set of requirements across two build options in this script. `pip install -r ...` is still in the script twice, once for each Objective-C and Python, but I think that's ok for now.

lidizheng commented on tools/internal_ci/helper_scripts/requirements.linux_perf.txt:
```diff
@@ -0,0 +1,5 @@
+scipy>=1.0.0
+requests>=2.19.0
+PyJWT>=2.0.1
+cryptography>=3.3.1,<4.0.0
```
nit: For CI dependencies, it's better to pin the package versions.

drfloob replied:
Done

lidizheng commented on tools/run_tests/python_utils/filter_pull_request_tests.py:
```diff
@@ -129,7 +129,7 @@ def _get_changed_files(base_branch):
                                 "HEAD"]).rstrip()
     return check_output(["git", "diff", base_commit, "--name-only",
-                         "HEAD"]).splitlines()
+                         "HEAD"]).decode("UTF-8").splitlines()
```
nit: Maybe we can update the `check_output` function itself? Then we don't need `decode()` in other places.

drfloob replied:
We can't, this is python's subprocess.check_output. For whatever reason, it was imported as `from subprocess import check_output`. I've changed the import to make that clear.

lidizheng reviewed (approved):
LGTM. Mostly minor comments.

jtattermusch reviewed (approved):
LGTM.

jtattermusch commented on tools/internal_ci/helper_scripts/prepare_build_macos_rc:
```diff
@@ -67,7 +68,7 @@ then
 
   # Needed for ios-binary-size
-  time pip install --user pyyaml pyjwt==1.7.1 pyOpenSSL cryptography requests
+  time pip install --user -r $DIR/requirements.macos.txt
```
since the "requirements.macos.txt" is only required for some jobs (AFAIK it's only needed for ios-binary-size),
should we further qualify the name of the requirements.macos.txt file  (just as you did for `requirements.linux_perf.txt`) to make it clear in what situations dependencies are actually needed?

drfloob replied:
`prepare_build_macos_rc` is currently sourced directly for a handful of tests:

```
./tools/internal_ci/macos/grpc_distribtests.sh:23:source tools/internal_ci/helper_scripts/prepare_build_macos_rc
./tools/internal_ci/macos/grpc_interop_toprod.sh:21:source tools/internal_ci/helper_scripts/prepare_build_macos_rc
./tools/internal_ci/macos/grpc_interop_toprod.sh:22:source tools/internal_ci/helper_scripts/prepare_build_macos_interop_rc
./tools/internal_ci/macos/grpc_run_tests_matrix.sh:21:source tools/internal_ci/helper_scripts/prepare_build_macos_rc
./tools/internal_ci/macos/grpc_build_artifacts.sh:25:source tools/internal_ci/helper_scripts/prepare_build_macos_rc
./tools/internal_ci/macos/grpc_run_bazel_c_cpp_tests.sh:22:source tools/internal_ci/helper_scripts/prepare_build_macos_rc
./tools/internal_ci/macos/grpc_ios_binary_size.sh:25:source tools/internal_ci/helper_scripts/prepare_build_macos_rc
```

I haven't gone any deeper to see how these dependencies are actually used in all of these tests. If they aren't needed, we could restructure the setup scripts as a future improvement, and make these requirements specific to the binary size test.

These dependencies are only used for Obj-C and Python ... we could rename the file `requirements.macos_objc_python.txt` to be more specific, if that's meaningful here. What do you think?

jtattermusch replied:
Let's leave as is. I was mostly confused about the "tabulate" and "scipy" dependencies

jtattermusch commented on tools/internal_ci/helper_scripts/requirements.macos.txt:
```diff
@@ -0,0 +1,7 @@
+pyOpenSSL==20.0.1
+PyYAML==5.4.1
+requests==2.25.1
+scipy==1.5.4
```
is scipy and tabulate actually needed on macOS?
I think you need those for bm_diff.py (which doesn't run on macos AFAIK) - can you please doublecheck?

drfloob replied:
They aren't needed, I'll remove them. I was toying with the idea of using one superset of all python requirements for all tests to simplify the build file structure, but decided against it since it might be needlessly costly (~30mb extra scipy download for every test, for example).

jtattermusch replied:
Now the list looks more reasonable, thanks!

drfloob commented:
Waiting on #25647 to fix our tests' pip installations.

jtattermusch commented:
Interop tests failures are https://github.com/grpc/grpc/issues/25652.
